modules/drivers/redshift/src/metabase/driver/redshift.clj

```diff
@@ -165,7 +165,24 @@
 (defn query->field-values
   "Convert a MBQL query to a map of field->value"
   [query]
-  (into {} (map (fn [param] [(:name (qp.store/field (get-in param [:target 1 1]))) (:value param)]) (:user-parameters query))))
+  (into {} (filter identity
+                   (map (fn [param]
+                          (cond
+                            ; if the parameter has a name already, use it
+                            (contains? param :name)
+                            [(:name param) (:value param)]
+
+                            ; if it's a dimension parameter with a "field-id", look up the field-id
+                            ; to get its name
+                            (= :dimension (first (:target param)))
+                            (let [dimension-params (apply array-map (second (:target param)))
+                                  field-id (:field-id dimension-params)]
+                              (when field-id
```

Comment:
I feel like there may be a better way to do this... My initial thought was `(when-let)` but that only allows a single binding, so this is what I came up with

Reply:
You could always do

```clj
(let [dimension-params ...]
  (when-let [field-id (:field-id dimension-params)]
    ...)
```

actually since you're not using `dimension-params` you could just do

```clj
(when-let [field-id (:field-id (apply array-map ...))]
  ...)
```

Reply:
We have a ton of functions for extracting stuff out of MBQL clauses. `metabase.mbql.util/match-one` would probably do what you want here:

```clj
(require '[metabase.mbql.util :as mbql.u])

(mbql.u/match-one [:dimension [:field-id 100]]
  [:field-id id]
  (when (= :dimension (first &parents))
    id))
;; -> 100
```

`mbql.u/match` and `mbql.u/match-one` are built on top of `clojure.core.match`, but they're recursive and bind a few anaphors for convenience. Check out the docstrings for `match` and `match-one`, I wrote really long essays explaining how they work. But a brief explanation of the example above is:

Basic syntax for usage is 

```clj
(mbql.u/match-one what-to-look-for-matches-in 
  match-pattern-1
  then-clause-1

  match-pattern-2
  then-clause-2

  ...)
``` 

1. We look for matches in `[:dimension [:field-id 100]]` 
1.  In this case we have one pair of `match-pattern`/`then-clause` --  `[:field-id id]` is the `match-pattern` and `(when ...)` is the `then-clause` -- what to do when we find something that can unify with `[:field-id id]` 
1.  Any vector with `:field-id` and any other arg will unify with `[:field-id id]` and bind the second arg to `id` 
1.  `match` and `match-one` will recursively iterate thru collections looking for matches until something unifies with a match pattern *and* its `then-clause` returns non-`nil` -- so you can put additional checks in the `then-clause` so it will continue iterating
1.  When recursing thru maps or MBQL clause vectors the path taken to get there is bound to the anaphor `&parents`. So because `:dimension` is the parent of `:field-id`, `&parents` will be bound to `[:dimension]` in this case

So you could write the example above like this and get the same behavior:

```clj
(mbql.u/match-one [:dimension [:field-id 100]]
  :dimension
  (mbql.u/match-one &match
    [:field-id id]
    id))
;; -> 100
```

You could even match against the whole map:

```clj
(mbql.u/match-one {:target  [:dimension [:field-id 100]]
                   :another [:dimension [:field-id 200]]}
  [:field-id id]
  (when (= [:target :dimension] (take-last 2 &parents))
    id))
;; -> 100
```

Reply:
After thinking about this a bit I think there's a few other cases we should handle as well. We support an `fk->` syntactic sugar clause that can be used in `:dimension` like this:

```clj
{:target [:dimension [:fk-> [:field-id 100] [:field-id 200]]]}
```

The `fk->` clause is syntactic sugar that tells the query processor to implicitly join another Table -- in this case, it tells us to use Field 100 (a foreign key) in the current source Table to join the Field 200's Table.

For example:

```sql
CREATE TABLE table_1 (field_100 INTEGER REFERENCES table_2.id);
CREATE TABLE table_2 (id INTEGER, field_200 text);
```

If we queried `table_1` and wanted to include `table_2.text` in the results we could use an `fk->` form like the one above to implicitly join `table_2.

Normally `fk->` is compiled out by the add-implicit-joins` middleware, and `metabase.driver.sql.query-processor` and the like never see it. But since that stage doesn't happen until after we handle parameters I suppose you might still run in to it here.

For dealing with different types of field clauses you can use `mbql.u/field-clause->id-or-literal`. This will return the ID of the Field that's actually being referred to (200 in the example above) or the name of the field for `field-literal` clauses (used to refer to fields in native queries). You can use this function to extract the ID or literal name and then check if it's an integer and return that:

```clj
(mbql.u/match-one [:dimension [:fk-> [:field-id 100] [:field-id 200]]]
  [:dimension field]
  (let [field-id-or-name (mbql.u/field-clause->id-or-literal field)]
    (when (integer? field-id-or-name)
      field-id-or-name)))
;; -> 200
```

Reply:
The various types of Field clauses are are little hairy, at some point in the future we'll hopefully switch to something more uniform with less variations but for now you can use the `mbql.u/` functions to make them easier to work with